File: oxidize-golang/core/mesh/ring_collective.go
package mesh

import (
	"context"
	"encoding/binary"
	"fmt"
	"math"
)

// RingError mirrors RingError. It captures the failure modes of a ring
// collective operation.
type RingError struct {
	Kind    string
	Message string
}

func (e *RingError) Error() string {
	if e.Message != "" {
		return "ring: " + e.Kind + ": " + e.Message
	}
	return "ring: " + e.Kind
}

func errWrongChunkSize(expected, actual int) *RingError {
	return &RingError{Kind: "wrong_chunk_size", Message: fmt.Sprintf("expected multiple of %d, got remainder %d", expected, actual)}
}

func errByteLengthMismatch(expected, actual int) *RingError {
	return &RingError{Kind: "byte_length_mismatch", Message: fmt.Sprintf("expected %d bytes, got %d", expected, actual)}
}

// RingNodeTransport is the abstract ring transport: each rank sends to its
// right neighbour and receives from its left neighbour. It mirrors the Rust
// RingTransport trait. Context cancellation aborts a blocked send/recv.
type RingNodeTransport interface {
	SendToRight(ctx context.Context, data []byte) error
	RecvFromLeft(ctx context.Context) ([]byte, error)
}

// RingBackend mirrors RingBackend. It performs ring all-reduce (AllSum) and
// ring all-gather (AllGather) over a RingNodeTransport.
type RingBackend struct {
	Rank      int
	NumRanks  int
	Transport RingNodeTransport
}

// NewRingBackend constructs a ring backend for a given rank.
func NewRingBackend(rank, numRanks int, transport RingNodeTransport) *RingBackend {
	return &RingBackend{Rank: rank, NumRanks: numRanks, Transport: transport}
}

// AllSum performs ring all-reduce, summing data in-place so every rank ends up
// with the same element-wise sum. len(data) must be divisible by NumRanks.
func (r *RingBackend) AllSum(ctx context.Context, data []float32) error {
	if r.NumRanks <= 1 {
		return nil
	}
	if len(data)%r.NumRanks != 0 {
		return errWrongChunkSize(r.NumRanks, len(data)%r.NumRanks)
	}

	chunkSize := len(data) / r.NumRanks
	sendBuf := make([]float32, chunkSize)
	recvF32 := make([]float32, chunkSize)

	// Scatter-reduce: N-1 steps.
	sendChunk := r.Rank
	recvChunk := (r.Rank + r.NumRanks - 1) % r.NumRanks
	for step := 0; step < r.NumRanks-1; step++ {
		sendOff := sendChunk * chunkSize
		copy(sendBuf, data[sendOff:sendOff+chunkSize])
		recvBytes, err := r.exchange(ctx, f32SliceToBytes(sendBuf))
		if err != nil {
			return err
		}
		if err := bytesToF32SliceInto(recvBytes, recvF32); err != nil {
			return err
		}
		recvOff := recvChunk * chunkSize
		for i := 0; i < chunkSize; i++ {
			data[recvOff+i] += recvF32[i]
		}
		sendChunk = (sendChunk + r.NumRanks - 1) % r.NumRanks
		recvChunk = (recvChunk + r.NumRanks - 1) % r.NumRanks
	}

	// All-gather: N-1 steps.
	sendChunk = (r.Rank + 1) % r.NumRanks
	recvChunk = r.Rank
	for step := 0; step < r.NumRanks-1; step++ {
		sendOff := sendChunk * chunkSize
		copy(sendBuf, data[sendOff:sendOff+chunkSize])
		recvBytes, err := r.exchange(ctx, f32SliceToBytes(sendBuf))
		if err != nil {
			return err
		}
		if err := bytesToF32SliceInto(recvBytes, recvF32); err != nil {
			return err
		}
		recvOff := recvChunk * chunkSize
		copy(data[recvOff:recvOff+chunkSize], recvF32)
		sendChunk = (sendChunk + r.NumRanks - 1) % r.NumRanks
		recvChunk = (recvChunk + r.NumRanks - 1) % r.NumRanks
	}

	return nil
}

// AllGather performs ring all-gather: each rank contributes one chunk (data)
// and receives the concatenation of all chunks into out. len(out) must equal
// len(data) * NumRanks.
func (r *RingBackend) AllGather(ctx context.Context, data []float32, out []float32) error {
	if r.NumRanks <= 1 {
		copy(out, data)
		return nil
	}
	chunkSize := len(data)
	if len(out) != chunkSize*r.NumRanks {
		return errWrongChunkSize(chunkSize*r.NumRanks, len(out))
	}

	sendBuf := make([]float32, chunkSize)
	recvF32 := make([]float32, chunkSize)

	localOffset := r.Rank * chunkSize
	copy(out[localOffset:localOffset+chunkSize], data)

	sendChunk := r.Rank
	recvChunk := (r.Rank + r.NumRanks - 1) % r.NumRanks
	for step := 0; step < r.NumRanks-1; step++ {
		sendOff := sendChunk * chunkSize
		copy(sendBuf, out[sendOff:sendOff+chunkSize])
		recvBytes, err := r.exchange(ctx, f32SliceToBytes(sendBuf))
		if err != nil {
			return err
		}
		if err := bytesToF32SliceInto(recvBytes, recvF32); err != nil {
			return err
		}
		recvOff := recvChunk * chunkSize
		copy(out[recvOff:recvOff+chunkSize], recvF32)
		sendChunk = (sendChunk + r.NumRanks - 1) % r.NumRanks
		recvChunk = (recvChunk + r.NumRanks - 1) % r.NumRanks
	}

	return nil
}

// exchange performs a concurrent send-to-right / recv-from-left, mirroring the
// Rust tokio::join! pattern so the ring does not deadlock.
func (r *RingBackend) exchange(ctx context.Context, send []byte) ([]byte, error) {
	type recvResult struct {
		data []byte
		err  error
	}
	recvCh := make(chan recvResult, 1)
	go func() {
		data, err := r.Transport.RecvFromLeft(ctx)
		recvCh <- recvResult{data, err}
	}()
	sendErr := r.Transport.SendToRight(ctx, send)
	res := <-recvCh
	if sendErr != nil {
		return nil, sendErr
	}
	if res.err != nil {
		return nil, res.err
	}
	return res.data, nil
}

// PipelineSend forwards activations to the next pipeline stage (right
// neighbour). Mirrors pipeline_send.
func PipelineSend(ctx context.Context, ring *RingBackend, activations []float32) error {
	return ring.Transport.SendToRight(ctx, f32SliceToBytes(activations))
}

// PipelineRecv receives activations from the previous pipeline stage (left
// neighbour). Mirrors pipeline_recv.
func PipelineRecv(ctx context.Context, ring *RingBackend, numFloats int) ([]float32, error) {
	bytes, err := ring.Transport.RecvFromLeft(ctx)
	if err != nil {
		return nil, err
	}
	out := make([]float32, numFloats)
	if err := bytesToF32SliceInto(bytes, out); err != nil {
		return nil, err
	}
	return out, nil
}

// TensorParallelAllSum performs a tensor-parallel all_sum over the ring.
// Mirrors tensor_parallel_all_sum.
func TensorParallelAllSum(ctx context.Context, ring *RingBackend, partial []float32) error {
	return ring.AllSum(ctx, partial)
}

// TensorParallelAllGather gathers outputs from all ranks. Mirrors
// tensor_parallel_all_gather.
func TensorParallelAllGather(ctx context.Context, ring *RingBackend, partial, out []float32) error {
	return ring.AllGather(ctx, partial, out)
}

// f32SliceToBytes converts a slice of float32 into little-endian bytes,
// matching the Rust framing.
func f32SliceToBytes(data []float32) []byte {
	out := make([]byte, len(data)*4)
	for i, v := range data {
		binary.LittleEndian.PutUint32(out[i*4:], math.Float32bits(v))
	}
	return out
}

// bytesToF32SliceInto converts little-endian bytes into a pre-allocated
// float32 slice.
func bytesToF32SliceInto(bytes []byte, out []float32) error {
	if len(bytes) != len(out)*4 {
		return errByteLengthMismatch(len(out)*4, len(bytes))
	}
	for i := range out {
		out[i] = math.Float32frombits(binary.LittleEndian.Uint32(bytes[i*4:]))
	}
	return nil
}

// bytesToF32Slice converts little-endian bytes into a freshly allocated
// float32 slice.
func bytesToF32Slice(bytes []byte) ([]float32, error) {
	if len(bytes)%4 != 0 {
		return nil, errByteLengthMismatch((len(bytes)/4+1)*4, len(bytes))
	}
	out := make([]float32, len(bytes)/4)
	if err := bytesToF32SliceInto(bytes, out); err != nil {
		return nil, err
	}
	return out, nil
}

// ChannelRingTransport is an in-memory ring transport used for tests, mirroring
// the Rust ChannelTransport.
type ChannelRingTransport struct {
	RightTx chan<- []byte
	LeftRx  <-chan []byte
}

// SendToRight pushes data onto the right neighbour's inbound channel.
func (c *ChannelRingTransport) SendToRight(ctx context.Context, data []byte) error {
	select {
	case c.RightTx <- data:
		return nil
	case <-ctx.Done():
		return &RingError{Kind: "io", Message: ctx.Err().Error()}
	}
}

// RecvFromLeft pops data from the left neighbour's channel.
func (c *ChannelRingTransport) RecvFromLeft(ctx context.Context) ([]byte, error) {
	select {
	case data, ok := <-c.LeftRx:
		if !ok {
			return nil, &RingError{Kind: "io", Message: "channel closed"}
		}
		return data, nil
	case <-ctx.Done():
		return nil, &RingError{Kind: "io", Message: ctx.Err().Error()}
	}
}

// CreateMockRing builds a mock ring of numRanks backends connected via
// channels. Mirrors create_mock_ring.
func CreateMockRing(numRanks int) []*RingBackend {
	channels := make([]chan []byte, numRanks)
	for i := range channels {
		channels[i] = make(chan []byte, 64)
	}
	backends := make([]*RingBackend, numRanks)
	for rank := 0; rank < numRanks; rank++ {
		leftRank := (rank + numRanks - 1) % numRanks
		transport := &ChannelRingTransport{
			// rank sends to its right neighbour's inbound channel.
			RightTx: channels[(rank+1)%numRanks],
			// rank receives from its own inbound channel (fed by its left).
			LeftRx: channels[rank],
		}
		_ = leftRank
		backends[rank] = NewRingBackend(rank, numRanks, transport)
	}
	return backends
}

File: oxidize-python/oxidize_python/core/mesh/ring_collective.py
"""Ring collective operations mirroring oxidize-golang/core/mesh/ring_collective.go.

Implements ring all-reduce (``all_sum``) and ring all-gather (``all_gather``)
over an abstract ring transport, plus pipeline / tensor-parallel helpers and
little-endian f32 framing matching the Rust/Go wire format.
"""

from __future__ import annotations

import queue
import struct
import threading
from abc import ABC, abstractmethod


class RingError(Exception):
    """Failure modes of a ring collective operation (mirrors RingError)."""

    def __init__(self, kind: str, message: str = "") -> None:
        self.kind = kind
        self.message = message
        if message:
            super().__init__(f"ring: {kind}: {message}")
        else:
            super().__init__(f"ring: {kind}")


def _err_wrong_chunk_size(expected: int, actual: int) -> RingError:
    return RingError(
        "wrong_chunk_size",
        f"expected multiple of {expected}, got remainder {actual}",
    )


def _err_byte_length_mismatch(expected: int, actual: int) -> RingError:
    return RingError("byte_length_mismatch", f"expected {expected} bytes, got {actual}")


class RingNodeTransport(ABC):
    """Abstract ring transport (mirrors the Rust RingTransport trait).

    Each rank sends to its right neighbour and receives from its left
    neighbour.
    """

    @abstractmethod
    def send_to_right(self, data: bytes) -> None: ...

    @abstractmethod
    def recv_from_left(self) -> bytes: ...


class RingBackend:
    """Ring all-reduce / all-gather backend (mirrors RingBackend)."""

    def __init__(self, rank: int, num_ranks: int, transport: RingNodeTransport) -> None:
        self.rank = rank
        self.num_ranks = num_ranks
        self.transport = transport

    def all_sum(self, data: list[float]) -> None:
        """Ring all-reduce, summing ``data`` in place.

        ``len(data)`` must be divisible by ``num_ranks``.
        """
        if self.num_ranks <= 1:
            return
        if len(data) % self.num_ranks != 0:
            raise _err_wrong_chunk_size(self.num_ranks, len(data) % self.num_ranks)

        chunk_size = len(data) // self.num_ranks

        # Scatter-reduce: N-1 steps.
        send_chunk = self.rank
        recv_chunk = (self.rank + self.num_ranks - 1) % self.num_ranks
        for _ in range(self.num_ranks - 1):
            send_off = send_chunk * chunk_size
            send_buf = data[send_off : send_off + chunk_size]
            recv_bytes = self._exchange(f32_to_bytes(send_buf))
            recv_f32 = bytes_to_f32(recv_bytes, chunk_size)
            recv_off = recv_chunk * chunk_size
            for i in range(chunk_size):
                data[recv_off + i] += recv_f32[i]
            send_chunk = (send_chunk + self.num_ranks - 1) % self.num_ranks
            recv_chunk = (recv_chunk + self.num_ranks - 1) % self.num_ranks

        # All-gather: N-1 steps.
        send_chunk = (self.rank + 1) % self.num_ranks
        recv_chunk = self.rank
        for _ in range(self.num_ranks - 1):
            send_off = send_chunk * chunk_size
            send_buf = data[send_off : send_off + chunk_size]
            recv_bytes = self._exchange(f32_to_bytes(send_buf))
            recv_f32 = bytes_to_f32(recv_bytes, chunk_size)
            recv_off = recv_chunk * chunk_size
            data[recv_off : recv_off + chunk_size] = recv_f32
            send_chunk = (send_chunk + self.num_ranks - 1) % self.num_ranks
            recv_chunk = (recv_chunk + self.num_ranks - 1) % self.num_ranks

    def all_gather(self, data: list[float]) -> list[float]:
        """Ring all-gather: contribute ``data`` and receive all chunks concatenated."""
        chunk_size = len(data)
        if self.num_ranks <= 1:
            return list(data)

        out = [0.0] * (chunk_size * self.num_ranks)
        local_offset = self.rank * chunk_size
        out[local_offset : local_offset + chunk_size] = data

        send_chunk = self.rank
        recv_chunk = (self.rank + self.num_ranks - 1) % self.num_ranks
        for _ in range(self.num_ranks - 1):
            send_off = send_chunk * chunk_size
            send_buf = out[send_off : send_off + chunk_size]
            recv_bytes = self._exchange(f32_to_bytes(send_buf))
            recv_f32 = bytes_to_f32(recv_bytes, chunk_size)
            recv_off = recv_chunk * chunk_size
            out[recv_off : recv_off + chunk_size] = recv_f32
            send_chunk = (send_chunk + self.num_ranks - 1) % self.num_ranks
            recv_chunk = (recv_chunk + self.num_ranks - 1) % self.num_ranks

        return out

    def _exchange(self, send: bytes) -> bytes:
        """Concurrent send-to-right / recv-from-left to avoid ring deadlock.

        Mirrors the Rust ``tokio::join!`` pattern with a receiver thread.
        """
        result: dict[str, object] = {}

        def _recv() -> None:
            try:
                result["data"] = self.transport.recv_from_left()
            except Exception as exc:  # noqa: BLE001
                result["err"] = exc

        recv_thread = threading.Thread(target=_recv)
        recv_thread.start()
        send_err: Exception | None = None
        try:
            self.transport.send_to_right(send)
        except Exception as exc:  # noqa: BLE001
            send_err = exc
        recv_thread.join()
        if send_err is not None:
            raise send_err
        if "err" in result:
            raise result["err"]  # type: ignore[misc]
        return result["data"]  # type: ignore[return-value]


def pipeline_send(ring: RingBackend, activations: list[float]) -> None:
    """Forward activations to the next pipeline stage (mirrors pipeline_send)."""
    ring.transport.send_to_right(f32_to_bytes(activations))


def pipeline_recv(ring: RingBackend, num_floats: int) -> list[float]:
    """Receive activations from the previous pipeline stage (mirrors pipeline_recv)."""
    data = ring.transport.recv_from_left()
    return bytes_to_f32(data, num_floats)


def tensor_parallel_all_sum(ring: RingBackend, partial: list[float]) -> None:
    """Tensor-parallel all_sum over the ring (mirrors tensor_parallel_all_sum)."""
    ring.all_sum(partial)


def tensor_parallel_all_gather(ring: RingBackend, partial: list[float]) -> list[float]:
    """Gather outputs from all ranks (mirrors tensor_parallel_all_gather)."""
    return ring.all_gather(partial)


def f32_to_bytes(data: list[float]) -> bytes:
    """Convert a list of float32 into little-endian bytes (matches Rust framing)."""
    return struct.pack(f"<{len(data)}f", *data)


def bytes_to_f32(data: bytes, count: int) -> list[float]:
    """Convert little-endian bytes into a list of ``count`` float32 values."""
    if len(data) != count * 4:
        raise _err_byte_length_mismatch(count * 4, len(data))
    return list(struct.unpack(f"<{count}f", data))


def bytes_to_f32_slice(data: bytes) -> list[float]:
    """Convert little-endian bytes into a freshly sized float32 list."""
    if len(data) % 4 != 0:
        raise _err_byte_length_mismatch((len(data) // 4 + 1) * 4, len(data))
    return bytes_to_f32(data, len(data) // 4)


class ChannelRingTransport(RingNodeTransport):
    """In-memory ring transport for tests (mirrors ChannelRingTransport)."""

    def __init__(self, right_tx: "queue.Queue[bytes]", left_rx: "queue.Queue[bytes]") -> None:
        self.right_tx = right_tx
        self.left_rx = left_rx

    def send_to_right(self, data: bytes) -> None:
        self.right_tx.put(data)

    def recv_from_left(self) -> bytes:
        return self.left_rx.get()


def create_mock_ring(num_ranks: int) -> list[RingBackend]:
    """Build a mock ring of ``num_ranks`` backends via in-memory channels.

    Mirrors create_mock_ring.
    """
    channels: list[queue.Queue[bytes]] = [queue.Queue(maxsize=64) for _ in range(num_ranks)]
    backends: list[RingBackend] = []
    for rank in range(num_ranks):
        transport = ChannelRingTransport(
            # rank sends to its right neighbour's inbound channel.
            right_tx=channels[(rank + 1) % num_ranks],
            # rank receives from its own inbound channel (fed by its left).
            left_rx=channels[rank],
        )
        backends.append(RingBackend(rank, num_ranks, transport))
    return backends
